### PR TITLE
Initial PR for multi-queue setup

### PR DESCRIPTION
**Pull request by nishkrishnan**
## Glossary
Merged Deploy - a deploy that is triggered from a new revision merging to the default branch (aka blessed deploy)
Manual Deploy - a deploy triggered from typing `atlantis apply -f` in a given PR.

## Problem

Currently we have two workflows:

Deploy Workflow 
* listens as new revisions are merged on the default branch and adds them to the queue.
* listens for `atlantis apply -f` comments and adds those revisions to the queue
* works off the queue 1 by 1, validates the revision for deployment and then calls Terraform workflow for that revision

Terraform Workflow
* Runs terraform plan and apply, prompting for approval in between

Using a single global queue in its current state doesn't allow us to lock deploy types independently.  We need to block merged deploy types from happening in the event of a manual deploy until the user confirms it's ok to do so.  This is similar to the existing code deploys world.  

## Approaches

Option 1: Priority Queue
We could theoretically change our queue to a priority to prioritize manual deploys.  We would still need to lock the queue only for merged deploys.  Currently, we sleep until the queue is empty which is an inexpensive check that gets invoked for every temporal state transition.  We would probably need to modify this condition and instead poll the queue.  This changes things from an event driven setup. 

Pros
* Likely would result in simpler high level setup with linear debugging

Cons
* Polling setup will likely increase our event history size.
* Increased complexity within our deploy workflow making code a bit harder to reason about.

Option 2:  3rd Proxy Workflow
We could have a 3rd workflow that is in between Terraform and Deploy that handles queuing up the Terraform workflow.  This workflow gets called directly when a revision is received that is a manual deploy.  It also gets called as each merged deploy is popped off the queue.  This makes it easy to just block the queue worker loop when there is a manual deploy.  

Pros
* Cleaner setup
* Code is easier to reason about

Cons
* Could result in annoyances tracing through 3 workflows.

This PR attempts to implement Option 2.

### CODE REVIEW

File: server/neptune/workflows/internal/deploy/revision/queue/worker.go [hunk 1]
```diff
@@ -1,18 +1,49 @@
...
+	LockedRootChecksSummary = "The current root has been locked. This is likely the result of a manual deployment. To override that lock and allow the main branch to perform new deployments, select the Unlock button."
+)
+
+func NewLockState() *LockState {
+	return &LockState{
+		status: Unlocked,
+	}
+}
+
+type LockState struct {
```
Comment (Aayyush) [inline server/neptune/workflows/internal/deploy/revision/queue/worker.go:33]: Thoughts on including `Lock` in the queue itself and exposing these methods on the lock: 

- `Lock()`
-  `Unlock()`
-  `GetLockStatus()`

Then, we wouldn't have to pass queue and lock state separately unless we plan to use `LockState` on it's own!

**Review by Aayyush — COMMENTED** [review]
Couple of thoughts:
- Unless the root is locked (after an FA), the queue will essentially be empty at all times since every deployment gets queued as signals in the proxy workflow. From what I remember, we originally planned to use a queue to have visibility into the queued deployments, which is not gonna be the case since the queue will be empty. Temporal UI should probably have a good UI on viewing queued signals but just wanted to point it out.

-  Regarding using Approach 1 (Priority Queue), we should still be able to sleep until the queue is empty right? We only need to pop the deployment with highest priority (latest FA request) and push to the queue in the right order. So, if FAs get queued up, we'll pop them in the order we received and all of them should be deployed as normal. Once we are done deploying all FAs, we pop the Merge Deploys which should hit `DirectionDirverged` and wait for a user to unlock before we can proceed with normal deploys. We'll probably have to check if the queue has any pending FA while waiting for the unlock signal and prioritize FA. Maybe I’m missing something that this approach does not handle, but just wanted to point it out.

**Comment by nishkrishnan** [comment]
> 



> Couple of thoughts:
> 
> * Unless the root is locked (after an FA), the queue will essentially be empty at all times since every deployment gets queued as signals in the proxy workflow. From what I remember, we originally planned to use a queue to have visibility into the queued deployments, which is not gonna be the case since the queue will be empty. Temporal UI should probably have a good UI on viewing queued signals but just wanted to point it out.

yeah that's a good point.  I also had the queue so that we could act on new revisions independently (ie. update the check run with queued status and/or scan the whole queue and perform some action on a given signal) which we can still do.

> * Regarding using Approach 1 (Priority Queue), we should still be able to sleep until the queue is empty right? We only need to pop the deployment with highest priority (latest FA request) and push to the queue in the right order. So, if FAs get queued up, we'll pop them in the order we received and all of them should be deployed as normal. Once we are done deploying all FAs, we pop the Merge Deploys which should hit `DirectionDirverged` and wait for a user to unlock before we can proceed with normal deploys. 

> We'll probably have to check if the queue has any pending FA while waiting for the unlock signal and prioritize FA. 
Yeah this is basically the annoyance, but I'm fine putting up another PR with this approach and we can compare. It's all new to me as well so hard to tell which one would be best long term.

**Comment by nishkrishnan** [comment]
>

**Pull request closed without merging**